File: daily/147.对链表进行插入排序.js
/*
 * @lc app=leetcode.cn id=147 lang=javascript
 *
 * [147] 对链表进行插入排序
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var insertionSortList = function(head) {
  if (!head) return head;
  let start = head;
  let end = head;
  while (end.next) {
    const cur = end.next;
    const next = cur.next;
    cur.next = null;
    end.next = null;
    let it = start;
    let pre = null;
    while (it && it.val < cur.val) {
      pre = it
      it = it.next;
    }
    if (pre) pre.next = cur;
    else start = cur;
    if (it) cur.next = it;
    else end = cur
    end.next = next;
  }
  return start;
};
// @lc code=end

const head = {
  val: -1,
  next: {
    val: 5,
    next: {
      val: 3,
      next: {
        val: 4,
        next: {
          val: 0,
          next: null
        }
      }
    }
  }
}

let a = insertionSortList(head);
while (a) {
  console.log(a.val)
  a = a.next;
}

File: daily/147.对链表进行插入排序.rs
/*
 * @lc app=leetcode.cn id=147 lang=rust
 *
 * [147] 对链表进行插入排序
 */

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}

// @lc code=start
// Definition for singly-linked list.

pub fn vec2list(mut arr: Vec<i32>) -> Option<Box<ListNode>> {
    match arr.pop() {
        Some(x) => Some(Box::new(ListNode {
            val: x,
            next: vec2list(arr),
        })),
        None => None,
    }
}

impl Solution {
    pub fn insertion_sort_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        if head.is_none() { return None; }
        let mut arr = vec![];
        let mut start = head;
        while let Some(x) = start {
            arr.push(x.val);
            start = x.next;
        }
        arr.sort_unstable_by(|a, b| b.cmp(&a));
        return vec2list(arr);
    }
}
// @lc code=end

